File: src/cryptoMM.js
/**
 * cryptoMM.js
 * Entry point for the Crypto Market Maker bot.
 * Posts two-sided quotes on Polymarket 5-minute BTC/ETH/SOL markets
 * to earn maker rebates + liquidity rewards.
 *
 * Run with: npm run cmm         (live)
 *           npm run cmm:paper   (paper trading with $1000 virtual balance)
 */

import config, { validateMMConfig } from './config/index.js';
import logger from './utils/logger.js';
import { initClientWithKeys } from './services/client.js';
import { startBinanceFeed, stopBinanceFeed, getBinanceFeedStatus } from './services/binanceFeed.js';
import { startSniperDetector, stopSniperDetector } from './services/sniperDetector.js';
import { startTimeframeDetector, stopTimeframeDetector } from './services/cryptoTimeframeDetector.js';
import { scheduleMarket, getMMStats, cancelAllOrders, isDailyLossHit, checkFills, CMM_ASSETS, loadSignalModels, saveState } from './services/cryptoMMExecutor.js';
import { sendCmmReport, sendHeartbeat, sendModelDegradationAlert, sendFeedStalenessAlert, ENABLED as TELEGRAM_ENABLED } from './services/telegram.js';
import { checkModelDegradation } from './services/cmmSignal.js';
import { checkAllFeedStaleness } from './services/binanceFeed.js';

// ── Validate ────────────────────────────────────────────────────────────────

validateMMConfig();

if (CMM_ASSETS.length === 0) {
    console.error('CMM_ASSETS is empty. Set e.g. CMM_ASSETS=btc,eth,sol in .env');
    process.exit(1);
}

const CMM_PRIVATE_KEY = config.tailSweepPrivateKey || config.privateKey;
const CMM_PROXY_WALLET = config.tailSweepProxyWallet || config.proxyWallet;

if (!CMM_PRIVATE_KEY || !CMM_PROXY_WALLET) {
    console.error('Missing wallet keys. Set TAILSWEEP_PRIVATE_KEY + TAILSWEEP_PROXY_WALLET_ADDRESS (or PRIVATE_KEY + PROXY_WALLET_ADDRESS) in .env');
    process.exit(1);
}

// ── Init CLOB client ────────────────────────────────────────────────────────

try {
    await initClientWithKeys(CMM_PRIVATE_KEY, CMM_PROXY_WALLET);
} catch (err) {
    logger.error(`CMM: Client init error: ${err.message}`);
    process.exit(1);
}

// ── Market handler ──────────────────────────────────────────────────────────

function handleNewMarket(market) {
    if (!CMM_ASSETS.includes(market.asset?.toLowerCase())) return;
    scheduleMarket(market);
}

// ── Status logging ──────────────────────────────────────────────────────────

let statusTimer = null;

function logStatus() {
    const stats = getMMStats();
    const feed = getBinanceFeedStatus();
    const mode = config.dryRun ? 'PAPER' : 'LIVE';
    const feedStatus = feed.status === 'connected' ? 'OK' : feed.status;
    const priceStr = feed.lastPrice ? `$${feed.lastPrice.toLocaleString()}` : 'N/A';

    const dailyPnlStr = stats.dailyPnl >= 0 ? `+$${stats.dailyPnl.toFixed(2)}` : `-$${Math.abs(stats.dailyPnl).toFixed(2)}`;
    const lossLimit = isDailyLossHit() ? ' [LOSS LIMIT HIT]' : '';

    logger.info(
        `CMM [${mode}] | active=${stats.activeMarkets} pending=${stats.pendingMarkets} | ` +
        `fills=${stats.fills} W=${stats.wins} L=${stats.losses} | ` +
        `daily=${dailyPnlStr}${lossLimit} | ` +
        `rewards~$${stats.dailyRewardEstimate.toFixed(2)} | ` +
        `feed=${feedStatus} BTC=${priceStr}`
    );
}

// ── Graceful shutdown ───────────────────────────────────────────────────────

async function shutdown() {
    logger.warn('CMM: shutting down...');
    stopSniperDetector();
    stopTimeframeDetector();
    stopBinanceFeed();
    await cancelAllOrders();
    if (fillTimer) clearInterval(fillTimer);
    if (statusTimer) clearInterval(statusTimer);

    const stats = getMMStats();
    const dailyPnlStr = stats.dailyPnl >= 0 ? `+$${stats.dailyPnl.toFixed(2)}` : `-$${Math.abs(stats.dailyPnl).toFixed(2)}`;
    logger.info(`CMM: final stats — fills=${stats.fills} W=${stats.wins} L=${stats.losses} daily=${dailyPnlStr}`);
    saveState();

    process.exit(0);
}

process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);

// ── Start ───────────────────────────────────────────────────────────────────

const CMM_TIMEFRAMES = (process.env.CMM_TIMEFRAMES || '15m,4h').split(',').map(s => s.trim().toLowerCase()).filter(Boolean);
const has5m = CMM_TIMEFRAMES.includes('5m');
const longTFs = CMM_TIMEFRAMES.filter(tf => tf !== '5m');

await loadSignalModels();

const mode = config.dryRun ? 'PAPER ($1000 virtual)' : 'LIVE';
logger.info(`CMM starting — ${mode}`);
logger.info(`Assets: ${CMM_ASSETS.join(', ').toUpperCase()} | Timeframes: ${CMM_TIMEFRAMES.join(', ')} | Spread: ${process.env.CMM_SPREAD || '0.04'} | Shares: ${process.env.CMM_SHARES || '20'} | Max daily loss: $${process.env.CMM_MAX_DAILY_LOSS || '50'}`);

// Cancel any stale open orders from previous sessions before posting new quotes
if (!config.dryRun) {
    try {
        const { getClient } = await import('./services/client.js');
        await getClient().cancelAll();
        logger.info('CMM: startup — stale orders cleared');
    } catch (err) {
        logger.warn(`CMM: startup cancelAll failed — ${err.message}`);
    }
}

// Start Binance feed for signal data
startBinanceFeed();

// 5-minute markets — only if '5m' is in CMM_TIMEFRAMES
// CMM_5M_ASSETS controls which assets run 5m (default: btc only)
if (has5m) {
    const assets5m = (process.env.CMM_5M_ASSETS || 'btc').split(',').map(s => s.trim().toLowerCase()).filter(Boolean);
    const origAssets = config.sniperAssets;
    config.sniperAssets = [...new Set([...origAssets, ...assets5m])];
    startSniperDetector(handleNewMarket);
    logger.info(`CMM: 5m detector started — assets: ${assets5m.join(', ').toUpperCase()}`);
}

// 1H+ markets
if (longTFs.length > 0) {
    startTimeframeDetector(longTFs, CMM_ASSETS, handleNewMarket);
    logger.info(`CMM: timeframe detector started — ${longTFs.join(', ')}`);
}

// Fill detection every 15 seconds — uses getTrades() for reliable fill tracking
const fillTimer = setInterval(async () => {
    try { await checkFills(); } catch (err) { logger.warn(`CMM: fill check error — ${err.message}`); }
}, 15_000);

// Status logging every 60 seconds
statusTimer = setInterval(logStatus, 60_000);
logStatus();

// Telegram: full report every 4h at 00/04/08/12/16/20 UTC
//           heartbeat (alive ping) every other hour — dead-man switch
const STATS_HOURS = new Set([0, 4, 8, 12, 16, 20]);
if (TELEGRAM_ENABLED) {
    const cmmMode = config.dryRun ? 'PAPER' : 'LIVE';

    function msUntilNextHour() {
        const now = new Date();
        const next = new Date(now);
        next.setUTCMinutes(0, 0, 0);
        next.setUTCHours(next.getUTCHours() + 1);
        return next.getTime() - now.getTime();
    }

    function scheduleNextHourlyTick() {
        setTimeout(async () => {
            const h = new Date().getUTCHours();
            const stats = getMMStats();
            if (STATS_HOURS.has(h)) {
                try { await sendCmmReport(stats, CMM_TIMEFRAMES); } catch {}
            } else {
                try { await sendHeartbeat(stats, cmmMode); } catch {}
            }

            // ML model degradation check (every hour)
            for (const asset of CMM_ASSETS) {
                const deg = checkModelDegradation(asset);
                if (deg.degraded) {
                    logger.warn(`CMM: ML model degraded for ${asset.toUpperCase()} — ${deg.reason}`);
                    try { await sendModelDegradationAlert(asset, deg.reason); } catch {}
                }
            }

            // Feed staleness check (every hour)
            const staleness = checkAllFeedStaleness();
            for (const [asset, result] of staleness) {
                if (result.stale) {
                    logger.warn(`CMM: feed stale for ${asset.toUpperCase()} — ${result.reason}`);
                    try { await sendFeedStalenessAlert(asset, result.staleDurationMs); } catch {}
                }
            }

            scheduleNextHourlyTick();
        }, msUntilNextHour());
    }

    scheduleNextHourlyTick();
    logger.info(`CMM: Telegram enabled — full report at 00/04/08/12/16/20 UTC, heartbeat every other hour`);
}

logger.info('CMM: waiting for markets...');

File: src/cryptoMM.py
"""
cryptoMM.py
Entry point for the Crypto Market Maker bot.
Posts two-sided quotes on Polymarket 5-minute BTC/ETH/SOL markets
to earn maker rebates + liquidity rewards.

Run with: python -m src.cryptoMM         (live)
          DRY_RUN=true python -m src.cryptoMM  (paper trading with $1000 virtual balance)
"""

import os
import sys
import signal
import asyncio
import datetime as dt

from src.config.index import config
from src.utils.logger import logger
from src.services.client import init_client_with_keys, get_client
from src.services.binanceFeed import start_binance_feed, stop_binance_feed, get_binance_feed_status
from src.services.sniperDetector import start_sniper_detector, stop_sniper_detector
from src.services.cryptoTimeframeDetector import start_timeframe_detector, stop_timeframe_detector
from src.services.cryptoMMExecutor import (
    schedule_market, get_mm_stats, cancel_all_orders,
    is_daily_loss_hit, check_fills, CMM_ASSETS
)
from src.services.telegram import send_cmm_report, ENABLED as TELEGRAM_ENABLED

# ── Market handler ──────────────────────────────────────────────────────────

def handle_new_market(market):
    asset = market.get('asset', '').lower()
    if asset not in CMM_ASSETS:
        return
    schedule_market(market)

# ── Status logging ──────────────────────────────────────────────────────────

def log_status():
    stats = get_mm_stats()
    feed = get_binance_feed_status()
    mode = "PAPER" if config.get('dryRun') else "LIVE"
    feed_status = 'OK' if feed.get('status') == 'connected' else feed.get('status')

    last_price = feed.get('lastPrice')
    price_str = f"${last_price:,}" if last_price else "N/A"

    daily_pnl = stats.get('dailyPnl', 0)
    if daily_pnl >= 0:
        daily_pnl_str = f"+${daily_pnl:.2f}"
    else:
        daily_pnl_str = f"-${abs(daily_pnl):.2f}"

    loss_limit = " [LOSS LIMIT HIT]" if is_daily_loss_hit() else ""
    daily_reward = stats.get('dailyRewardEstimate', 0)

    logger.info(
        f"CMM [{mode}] | active={stats.get('activeMarkets', 0)} pending={stats.get('pendingMarkets', 0)} | "
        f"fills={stats.get('fills', 0)} W={stats.get('wins', 0)} L={stats.get('losses', 0)} | "
        f"daily={daily_pnl_str}{loss_limit} | "
        f"rewards~${daily_reward:.2f} | "
        f"feed={feed_status} BTC={price_str}"
    )

# ── Graceful shutdown ───────────────────────────────────────────────────────

_shutting_down = False

async def shutdown(sig):
    global _shutting_down
    if _shutting_down:
        return
    _shutting_down = True

    logger.warning("CMM: shutting down...")
    stop_sniper_detector()
    stop_timeframe_detector()
    stop_binance_feed()
    await cancel_all_orders()

    stats = get_mm_stats()
    daily_pnl = stats.get('dailyPnl', 0)
    daily_pnl_str = f"+${daily_pnl:.2f}" if daily_pnl >= 0 else f"-${abs(daily_pnl):.2f}"
    logger.info(f"CMM: final stats — fills={stats.get('fills', 0)} W={stats.get('wins', 0)} L={stats.get('losses', 0)} daily={daily_pnl_str}")

    tasks = [t for t in asyncio.all_tasks() if t is not asyncio.current_task()]
    for task in tasks:
        task.cancel()
    await asyncio.gather(*tasks, return_exceptions=True)

# ── Main ────────────────────────────────────────────────────────────────────

async def main():
    # ── Validate ────────────────────────────────────────────────────────
    if not CMM_ASSETS:
        logger.error("CMM_ASSETS is empty. Set e.g. CMM_ASSETS=btc,eth,sol in .env")
        sys.exit(1)

    cmm_private_key = config.get('tailSweepPrivateKey') or config.get('privateKey')
    cmm_proxy_wallet = config.get('tailSweepProxyWallet') or config.get('proxyWallet')

    if not cmm_private_key or not cmm_proxy_wallet:
        logger.error(
            "Missing wallet keys. Set TAILSWEEP_PRIVATE_KEY + TAILSWEEP_PROXY_WALLET_ADDRESS "
            "(or PRIVATE_KEY + PROXY_WALLET_ADDRESS) in .env"
        )
        sys.exit(1)

    # ── Init CLOB client ────────────────────────────────────────────────
    try:
        await init_client_with_keys(cmm_private_key, cmm_proxy_wallet)
    except Exception as err:
        logger.error(f"CMM: Client init error: {err}")
        sys.exit(1)

    # ── Register signal handlers ────────────────────────────────────────
    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, lambda s=sig: asyncio.create_task(shutdown(s)))

    # ── Start ───────────────────────────────────────────────────────────
    cmm_timeframes_env = os.environ.get('CMM_TIMEFRAMES', '5m')
    cmm_timeframes = [s.strip().lower() for s in cmm_timeframes_env.split(',') if s.strip()]
    has_5m = '5m' in cmm_timeframes
    long_tfs = [tf for tf in cmm_timeframes if tf != '5m']

    mode = "PAPER ($1000 virtual)" if config.get('dryRun') else "LIVE"
    logger.info(f"CMM starting — {mode}")

    assets_str = ", ".join(CMM_ASSETS).upper()
    tfs_str = ", ".join(cmm_timeframes)
    spread = os.environ.get('CMM_SPREAD', '0.04')
    shares = os.environ.get('CMM_SHARES', '20')
    max_loss = os.environ.get('CMM_MAX_DAILY_LOSS', '50')
    logger.info(f"Assets: {assets_str} | Timeframes: {tfs_str} | Spread: {spread} | Shares: {shares} | Max daily loss: ${max_loss}")

    # Cancel any stale open orders from previous sessions
    if not config.get('dryRun'):
        try:
            client = get_client()
            await client.cancel_all()
            logger.info("CMM: startup — stale orders cleared")
        except Exception as err:
            logger.warning(f"CMM: startup cancelAll failed — {err}")

    # Start Binance feed for signal data
    start_binance_feed()

    # 5-minute markets
    if has_5m:
        orig_assets = config.get('sniperAssets', [])
        config['sniperAssets'] = list(set(orig_assets + CMM_ASSETS))
        start_sniper_detector(handle_new_market)
        logger.info("CMM: 5m detector started")

    # 1H+ markets
    if long_tfs:
        start_timeframe_detector(long_tfs, CMM_ASSETS, handle_new_market)
        logger.info(f"CMM: timeframe detector started — {', '.join(long_tfs)}")

    # Fill detection every 15 seconds
    async def fill_loop():
        while True:
            await asyncio.sleep(15)
            try:
                await check_fills()
            except Exception as err:
                logger.warning(f"CMM: fill check error — {err}")

    # Status logging every 60 seconds
    async def status_loop():
        while True:
            await asyncio.sleep(60)
            log_status()

    log_status()
    asyncio.create_task(fill_loop())
    asyncio.create_task(status_loop())

    # Telegram report 3x/day at 00:00, 08:00, 16:00 UTC
    if TELEGRAM_ENABLED:
        async def telegram_loop():
            while True:
                now = dt.datetime.now(dt.timezone.utc)
                h = now.hour
                next_hour = next((t for t in [0, 8, 16] if t > h), 24)

                next_time = now.replace(hour=next_hour % 24, minute=0, second=0, microsecond=0)
                if next_hour == 24:
                    next_time += dt.timedelta(days=1)

                wait_sec = (next_time - now).total_seconds()
                await asyncio.sleep(wait_sec)

                try:
                    await send_cmm_report(get_mm_stats(), cmm_timeframes)
                except Exception:
                    pass

        asyncio.create_task(telegram_loop())
        logger.info("CMM: Telegram reports enabled — next at next 00/08/16 UTC boundary")

    logger.info("CMM: waiting for markets...")

    # Keep alive
    try:
        while True:
            await asyncio.sleep(3600)
    except asyncio.CancelledError:
        pass

if __name__ == '__main__':
    asyncio.run(main())
